text/0000-binary-ops-specialization.md

```diff
@@ -135,93 +153,72 @@ where T: ?Sized + Borrow<str>
-the legacy traits, but new-style impls should be preferred in new APIs.
-
-The interleaved, specialized-first overload resolution rule is designed to
-prevent "spooky action at a distance" where e.g. adding a blanket impl of
-`Add2` for type `A` defined in one crate could shadow an existing
-`impl Add<B> for A` in another crate that defines `B`. The situation where
-non-generic impls of `Add` and `Add2`, defined in different crates, could
-apply to the same pair of types, is impossible due to acyclicity of crate
-dependencies and the orphan rule.
+
+Specialized implementations of the default overload traits can be defined
+when practical, within the `Borrow` bound of the default implementation.
+Crate authors are also free to provide new impls of the plain old overload
+traits, which override the generic impls of the default overload traits
+for purposes of operator overloading, or apply outside of the type families
+circumscribed by the default overload trait impls.
 
 # Drawbacks
 [drawbacks]: #drawbacks
 
-The second-generation traits add complexity, especially to operator
-overload resolution. It's likely that both new and old school trait impls
-will have to be provided side by side, which increases the possibility of
+The fallback overload traits add complexity, especially to operator
+overload resolution. It's likely that implementations for both default and
+plain old operator traits will have to be provided side by side to support
+older versions of the compiler, which increases the possibility of
 implementation errors. This takes further the precedent set by the
 [specialization RFC][rfc1210] that multiple different implementations may be
 considered to fit one use.
 
-`Borrow<T>` may prove too inflexible a bound for some interoperable
-type families. It's a challenge, though, to come up with an example in
-existing designs that goes beyond a few closely knit types.
-
 # Rationale and alternatives
 [rationale-and-alternatives]: #rationale-and-alternatives
 
-The proposed [rule][default-blanket-implementation-rule] of defining
-operator impls slashes the combinatorial explosion of mostly tedious
-operator trait implementations seen today, leaving reasonable flexibility
-in which operand type pairs are allowed (with `Borrow` as the guiding force).
-
-Addition of second-generation traits on top of the existing system
-provides a backward-compatible migration path for the Rust 1.x timeframe.
-
-If opt-in feature gates were possible in the stable channel, the new default
-impls could be defined for the Rust 1.0 traits and hidden behind a feature
-gate. It's unclear to the author if this could work without the need for
-all crates in the dependency graph to be compatible with the feature.
+The proposed [rule][default-implementation-rule] of defining generic
+operator impls slashes the quadratic explosion of mostly tedious
+non-generic operator trait implementations that takes place today.
```

Comment:
Most of this tedium could be resolved through disciplined use of macros.

Reply:
I'm [doing this](https://github.com/mzabaluev/strchunk/blob/f04dba281e867c4317f2ebc5a78c1df8fac8d6b6/src/impls.rs#L82), and it's still tedious :smile:

Reply:
The macros also don't help when there is another clever container crate whose types compare to the same basic types as yours. You'd have to add a dependency on the crate, just for the impls. The authors of that crate might want to do the same for your crate, and now you have to mutually agree over what depends on what and under which feature flags.

The generic impl does it automagically and with no extra dependencies, as long as there is a fitting `Borrow`.